R/ui.R

```diff
@@ -0,0 +1,132 @@
+  api_url <- getOption(
+    "plumber.apiURL",
+    urlHost(
+      scheme = getOption("plumber.apiScheme", "http"),
+      host   = getOption("plumber.apiHost", host),
+      port   = getOption("plumber.apiPort", port),
+      path   = getOption("plumber.apiPath", ""),
+      changeHostLocation = TRUE
+    )
+  )
+
+  # Mount openAPI spec paths openapi.json
+  mountOpenAPI(pr, api_url)
+
+  # Mount UIs
+  if (isTRUE(ui)) {
+    ui <- getOption("plumber.ui", "swagger")
+  }
+  interface <- ui[1]
+  ui_mount <- .globals$interfaces[[interface]]
+  if (!is.null(ui_mount)) {
+    ui_url <- ui_mount(pr, api_url, ...)
+    message("Running ", interface, " UI at ", ui_url, sep = "")
+  } else {
+    message("Ignored unknown user interface ", interface,". Supports ",
+            paste0('"', names(.globals$interfaces), '"', collapse = ", "))
+  }
+
+  # Use callback when defined
+  if (is.function(callback)) {
+    callback(ui_url)
+  }
+
+  return(invisible())
+
+}
+
+#' Mount OpenAPI Specification to a plumber router
+#' @noRd
+mountOpenAPI <- function(pr, api_url) {
+
+  spec <- pr$swaggerFile()
+
+  # Create a function that's hardcoded to return the OpenAPI specification -- regardless of env.
+  openapi_fun <- function(req) {
+    # use the HTTP_REFERER so RSC can find the swagger location to ask
+    ## (can't directly ask for 127.0.0.1)
+    if (is.null(getOption("plumber.apiURL")) &&
+        is.null(getOption("plumber.apiHost"))) {
+      if (is.null(req$HTTP_REFERER)) {
+        # Prevent leaking host and port if option is not set
+        api_url <- character(1)
+      }
+      else {
+        # Use HTTP_REFERER as fallback
+        api_url <- req$HTTP_REFERER
+        api_url <- sub("index\\.html$", "", api_url)
+        api_url <- sub(paste0("__(", paste0(names(.globals$interfaces), collapse = "|"),")__/$"), "", api_url)
+      }
+    }
+
+    utils::modifyList(list(servers = list(list(url = api_url))), spec)
+
+  }
+  # http://spec.openapis.org/oas/v3.0.3#document-structure
+  # "It is RECOMMENDED that the root OpenAPI document be named: openapi.json"
+  pr$handle("GET", "/openapi.json", openapi_fun, serializer = serializer_unboxed_json())
+
+  return(invisible())
+
+}
+
+#' Mount Interface UI
+#' @noRd
+mountInterfaces <- function(interface) {
+  mountInterface <- function(pr, api_url, ...) {
+    if (!requireNamespace(interface$package, quietly = TRUE)) {
+      stop(interface$package, " must be installed for the ", interface$name," UI to be displayed")
+    }
+
+    interfacePath <- paste0("/__", tolower(interface$name), "__/")
+    interfaceUrl <- paste0(api_url, interfacePath)
+
+    html_content <- interface$index(...)
+    interface_index <- function() {
+      html_content
+    }
+    for (path in paste0(interfacePath, c("index.html", ""))) {
+      pr$handle(
+        "GET", path, interface_index,
+        serializer = serializer_html()
+      )
+    }
+    pr$mount(interfacePath, PlumberStatic$new(interface$static()))
```

Comment:
Also changing `interface$static` to an object over a method.

```suggestion
    if (!is.null(interface$static)) {
      pr$mount(interfacePath, PlumberStatic$new(interface$static))
    }
```

Reply:
removing

Reply:
Why? Both swagger_path redoc_path do not accept arguments.

Reply:
I don't know if we should pass in the OAS `version`, `pr` (which is prolly not used), or nothing.

It is possible that down the line, my version of plumber has a different OAS version than someone else's version. However, both of us would be able to use the same version of `swagger` pkg. `swagger` would then retrieve the proper OAS ui matching the version.

Reply:
What I think
- `version` should come from plumber in the args list, I'll remove it from the interface.
- `pr` should not be passed as it would require `swagger` to be aware of the current `pr` in plumber

Reply:
Works for me!